packages/@lwc/template-compiler/src/parser/expression-complex/validate.ts

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { ParserDiagnostics, invariant } from '@lwc/errors';
+import { walk } from 'estree-walker';
+import * as t from '../../shared/estree';
+import type { BaseNode } from 'estree';
+
+const ALWAYS_INVALID_TYPES = new Map(
+    Object.entries({
+        AwaitExpression: 'await expressions',
+        ClassExpression: 'classes',
+        SequenceExpression: 'comma operators',
+        FunctionExpression: 'function expressions',
+        ImportExpression: 'imports',
+        MetaProperty: 'import.meta',
+        NewExpression: 'object instantiation',
+        RegExpLiteral: 'regular expression literals',
+        Super: '`super`',
+        ThisExpression: '`this`',
+        YieldExpression: '`yield`',
+    })
+);
+const STATEMENT_TYPES = new Set([
+    'Statement',
+    'BlockStatement',
+    'BreakStatement',
+    'ContinueStatement',
+    'DebuggerStatement',
+    'DoWhileStatement',
+    'EmptyStatement',
+    'ExpressionStatement',
+    'ForInStatement',
+    'ForStatement',
+    'FunctionDeclaration',
+    'IfStatement',
+    'LabeledStatement',
+    'ReturnStatement',
+    'SwitchStatement',
+    'ThrowStatement',
+    'TryStatement',
+    'VariableDeclaration',
+    'WhileStatement',
+    'WithStatement',
+    'ClassDeclaration',
+    'ExportAllDeclaration',
+    'ExportDefaultDeclaration',
+    'ExportNamedDeclaration',
+    'ForOfStatement',
+    'ImportDeclaration',
+    'DeclareClass',
+    'DeclareFunction',
+    'DeclareInterface',
+    'DeclareModule',
+    'DeclareModuleExports',
+    'DeclareTypeAlias',
+    'DeclareOpaqueType',
+    'DeclareVariable',
+    'DeclareExportDeclaration',
+    'DeclareExportAllDeclaration',
+]);
+const EARLY_STAGE_LANGUAGE_FEATURES = new Set([
+    'BindExpression',
+    'DecimalLiteral',
+    'DoExpression',
+    'RecordExpression',
+    'TupleExpression',
+    'ModuleExpression',
+    'TopicReference',
+    'PipelineTopicExpression',
+    'PipelineBareFunction',
+    'PipelinePrimaryTopicReference',
+]);
+const MUTATION_TYPES = new Set(['AssignmentExpression', 'UpdateExpression']);
+
+function validateArrowFunction(node: t.ArrowFunctionExpression) {
+    invariant(node.body.type !== 'BlockStatement', ParserDiagnostics.INVALID_EXPR_ARROW_FN_BODY);
+    invariant(!node.async, ParserDiagnostics.INVALID_EXPR_ARROW_FN_KIND, ['async']);
+    // This condition should never occur, unless the spec changes. However, it is
+    // permitted by the ESTree representation, so we'll check for it just in case.
+    invariant(!node.generator, ParserDiagnostics.INVALID_EXPR_ARROW_FN_KIND, ['generators']);
+}
+
+function validateUnaryExpression(node: t.UnaryExpression) {
+    invariant(node.operator !== 'delete', ParserDiagnostics.INVALID_EXPR_DELETE_OP);
+}
+
+function validateLiteral(node: t.Literal) {
```

Comment:
What is the reason behind restricting the usage of regular and big int literals?

Reply:
There was some concern that big ints wouldn't be supported in all browsers we ship to. Since there isn't a compelling reason to provide support for `BigInt`s, and because proper support in older browsers would probably include a polyfill, disallowing seemed prudent.

Regular expression literals were originally permitted in the design doc. However, they can be difficult to visually parse, and there wasn't a compelling use case, so it was moved to the prohibited column.

This was something I figured we could revisit if we get strong feedback from pilot users.